### DISCUSSION TITLE
lab D - banking question

### DISCUSSION THREAD

**Discussion by Josephood7** [bol-edu/HLS-SOC-Discussions#188] [Q&A] [answered]
對不起沒有搞得很懂
我的理解是有4個bank
1. 一個放data 一個放address 一個分配空間用來運算 一個拿來當fetch data的空間
2. 這裡的data是指data是指firmware code還是只要運算的adder.h裡定義的data
3. 還有想問code execution是指allocate空間來當作dram來執行運算嗎
好混亂 自己沒有搞得很懂這4個bank的分配 望大家指教
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/144367420/677c0793-a9cd-4e99-b87c-deddfc9688bb)
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/144367420/06c6c63c-79ad-4043-a402-ea2cd9b9c19f)

**Comment by ZONG-RUI1216** [comment]
我的理解是prefetch cache是額外開出來的記憶體空間(我是用register做)，與這四個bank沒有關係，助教應該只是希望你能至少分成data(.h定義裡的data) 與instructure (firmware execution code) 這兩種bank ，以我們組來說，因為instructure 用量較大，我就分配兩個bank給instructure ，所有的data分配剩下的兩個bank
那prefetch cache你就可以分成存data cache與instructure cache，去訪問不同的bank來得到你要的資訊

**Reply by Josephood7** [reply]
同學您(.h定義裡的data) 與(firmware execution code) 是指要在哪個部分去把他們分開呀

**Reply by ZONG-RUI1216** [reply]
修改sections.lds 裡的linkers address，首先要把(.h定義裡的data)從原本會加載到dff變成加載到mprjram(也就是0x38000000)中，
修改你可以發現A B等..矩陣從原本0x00000000開頭(如下圖)
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/145251222/c80e7169-b6fa-4289-b1c6-71513d6d36c6)
變成了0x38...開頭(如下圖)
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/145251222/a90828c6-df6f-465e-9a9d-a0b975167728)
接下來因為firmware execution code本身就是會存放在0x38...開頭中的address，那你就還需要一些方式去區分data與instructure ，舉例來說，我讓Bank address [9:8]這兩個bits的0 1都分配給instructure  (也就是說0x38000000-0x380001FF我都只能用來放instructure code) ，那A、B矩陣我放在Bank address [9:8] 3的位置(也就是說0x38000300-0x380003FF我都只能用來放A、B 矩陣的data)，至於怎麼修改可以參考下面那位同學的做法~

**Comment by kevin861222** [comment]
你可以先在終端機 下 "size" 指令，來看各 section 佔多少空間，截圖是 adder 的，共10筆型態為 int 的 data，一個 Bank 可以放的下。
![截圖 2023-12-26 下午10 39 30](https://github.com/bol-edu/HLS-SOC-Discussions/assets/79128379/01f63abc-faa3-41a9-82be-b51eda68a851)

另外 Code Execution 需要兩個 Bank 才能完全存放，我想這次實驗將資料分開至不同 Bank ，是為了方便地址判斷該資料意涵，方便我們決定要不要 prefetch。

Prefetch 的部分，workbook 中只有說明 'serve as cache' ，所以我也是使用 reg 代替，目的是模擬出 access time 只要 1T 的 memory

**Reply by Josephood7** [reply]
了解~ 但我不太懂linker那邊要改甚麼部分 是要改成儲存在sram嗎 還是說這邊也要把data跟instruction分開呢

**Reply by jxes993409** [reply, marked as answer]
同學你好 可以參考我這篇的[留言](https://github.com/bol-edu/HLS-SOC-Discussions/discussions/189#discussioncomment-7915925)看看

**Comment by boledulab** [comment]
SDRAM Bank selection is based on pre-determined bank addresses, e.g., A[9,8]. You can choose the bank address that coordinates with the linker.  
The prefetch buffer is more flexible and adapts to dynamic run time behavior. The number of prefetch buffers depends on how many data threads are accessed concurrently. For example, if you have three accelerators running simultaneously plus an instruction stream. Four prefetch buffers will provide the best result. Of course, it also needs to coordinate with the arbitrator.
